### PR TITLE
sanitize dir name before using it on odo push

### PR DESCRIPTION
**Pull request by girishramnani**
**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`

 /kind bug

**What does does this PR do / why we need it**:
sanitize dir name before using it on odo push

**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/4305

**PR acceptance criteria**:

- [X] Unit test 

- [ ] Integration test 

- [ ] Documentation 

- [X] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
- create a devfile with no `name` metadata field
-  do `odo push` on it with some invalid directory names
   - 4564564
   - staticHTML
   - 54sldkff%%23

And it would santize them before using them as names

### CODE REVIEW

File: pkg/util/util.go
Comment (dharmit) [inline pkg/util/util.go:319]: How does checking the length of variables help us determine if the name is all numeric? What am I missing here?

**Review by dharmit — CHANGES_REQUESTED** [review]
:+1: for removing an `else` block there. But I'm confused.

File: pkg/util/util.go
Comment (girishramnani) [reply pkg/util/util.go:319]: If the len of the original string is not 0 is the control condition just or make sure no one passed a blank directory,
Then after converting it to DNS1123 - if all are numeric the “removeNonAlphaprefix” would remove all characters as everything is non alpha hence returning a blank string.

**Comment by dharmit** [comment]
@girishramnani can you PTAL at the failing integration test for this PR?

File: pkg/url/url.go
Comment (girishramnani) [inline pkg/url/url.go:1122]: GetDNS1123Name now has a way stronger check - not needed for a url string

**Review by dharmit — COMMENTED** [review]
LGTM if url tests are happy. I'll add the lgtm label when I see the CI green or at least not red because of failing `odo url` tests.

**Comment by dharmit** [comment]
/retest

CI failing due to #4347

**Comment by girishramnani** [comment]
This PR would break existing urls

**Comment by dharmit** [comment]
> This PR would break existing urls

/hold

**Comment by girishramnani** [comment]
we can say that this one was bot's fault

**Comment by dharmit** [comment]
I hope it doesn't fire back. :crossed_fingers:

**Comment by kadel** [comment]
So we know that it can break existing users and leaving it as it is?? 😱

**Comment by dharmit** [comment]
> This PR would break existing urls

@girishramnani Can you elaborate a little as to where you think this could blow up?

**Pull request merged**
